File: languages/rust/fannkuch.rs
use std::env;

fn fannkuch(n: usize) -> (i32, i32) {
    let mut perm1: Vec<usize> = (0..n).collect();
    let mut perm = vec![0usize; n];
    let mut count = vec![0usize; n];
    let mut max_flips = 0i32;
    let mut checksum = 0i32;
    let mut perm_idx = 0i64;
    let mut r = n;

    loop {
        while r != 1 {
            count[r - 1] = r;
            r -= 1;
        }

        perm.copy_from_slice(&perm1);
        let mut flips = 0i32;
        let mut k = perm[0];
        while k != 0 {
            let (mut i, mut j) = (0usize, k);
            while i < j {
                perm.swap(i, j);
                i += 1;
                j -= 1;
            }
            flips += 1;
            k = perm[0];
        }

        if flips > max_flips {
            max_flips = flips;
        }
        if perm_idx % 2 == 0 {
            checksum += flips;
        } else {
            checksum -= flips;
        }

        // Generate the next permutation.
        loop {
            if r == n {
                return (max_flips, checksum);
            }
            let first = perm1[0];
            for i in 0..r {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = first;
            count[r] -= 1;
            if count[r] > 0 {
                break;
            }
            r += 1;
        }
        perm_idx += 1;
    }
}

fn main() {
    let n: usize = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(7);
    let (max_flips, checksum) = fannkuch(n);
    println!("{}", checksum);
    println!("Pfannkuchen({}) = {}", n, max_flips);
}

File: languages/swift/fannkuch.swift
import Foundation

func fannkuch(_ n: Int) -> (Int, Int) {
    var perm1 = Array(0..<n)
    var perm = Array(repeating: 0, count: n)
    var count = Array(repeating: 0, count: n)
    var maxFlips = 0
    var checksum = 0
    var permIdx = 0
    var r = n

    while true {
        while r != 1 {
            count[r - 1] = r
            r -= 1
        }

        perm = perm1
        var flips = 0
        perm.withUnsafeMutableBufferPointer { buf in
            var k = buf[0]
            while k != 0 {
                var i = 0
                var j = k
                while i < j {
                    let tmp = buf[i]
                    buf[i] = buf[j]
                    buf[j] = tmp
                    i += 1
                    j -= 1
                }
                flips += 1
                k = buf[0]
            }
        }

        if flips > maxFlips {
            maxFlips = flips
        }
        checksum += (permIdx % 2 == 0) ? flips : -flips

        // Generate the next permutation.
        while true {
            if r == n {
                return (maxFlips, checksum)
            }
            let first = perm1[0]
            for i in 0..<r {
                perm1[i] = perm1[i + 1]
            }
            perm1[r] = first
            count[r] -= 1
            if count[r] > 0 {
                break
            }
            r += 1
        }
        permIdx += 1
    }
}

let n = CommandLine.arguments.count > 1 ? (Int(CommandLine.arguments[1]) ?? 7) : 7
let (maxFlips, checksum) = fannkuch(n)
print(checksum)
print("Pfannkuchen(\(n)) = \(maxFlips)")
